### PR TITLE
UI tests to change password from user personal settings

### PR DESCRIPTION
**Pull request by sushilkshetri**
## Description
UI tests to change password from user personal settings

## Motivation and Context
This tests checks the password change on user setting page

## How Has This Been Tested?
Locally tested

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: tests/ui/features/lib/PersonalGeneralSettingsPage.php [hunk 1]
```diff
@@ -75,4 +78,17 @@ public function waitTillPageIsLoaded(
...
+	 *
+	 * @param string $oldPassword
+	 * @param string $newPassword
+	 *
+	 * @return void
+	 */
+	public function changePassword($oldPassword, $newPassword) {
+		$this->fillField($this->newPasswordInputID, $newPassword);
+		$this->fillField($this->oldPasswordInputID, $oldPassword);
+		$this->find('xpath', $this->xpathOfChangePasswordButton)->click();
```
Comment (individual-it) [inline tests/ui/features/lib/PersonalGeneralSettingsPage.php]: let's check if `find()` returns `null` as we do in other places before clicking
and we should be able to use `findById()` and not `find("xpath", ...)` to make it simpler

**Review by individual-it — CHANGES_REQUESTED** [review]
looks generally very good, just small change requests 👍

File: tests/ui/features/lib/PersonalGeneralSettingsPage.php (hunk 1 above)
Comment (phil-davis) [reply tests/ui/features/lib/PersonalGeneralSettingsPage.php]: Nice to see that a button has an id that can be used!

**Review by phil-davis — COMMENTED** [review]
Looks good. Just try the change suggested by Artur.

**Comment by sushilkshetri** [comment]
@phil-davis , @individual-it  please review again I have added some more tests.

File: tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php [hunk 2]
```diff
@@ -77,4 +82,56 @@ public function iChangeTheLanguageTo($language) {
...
+	
+	/**
+	 * @Then a message should be displayed with the text :wrongPasswordmessageText
+	 *
+	 * @param string $wrongPasswordmessageText
+	 *
+	 * @return void
+	 */
+	public function aMessageShouldBeDisplayedWithTheText($wrongPasswordmessageText) {
+		PHPUnit_Framework_Assert::assertEquals($wrongPasswordmessageText, $this->personalGeneralSettingsPage->getWrongPasswordMessageText($this->getSession()));
```
Comment (individual-it) [inline tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php]: that line is far too long. Please make it shorter. Check also the coding standards https://doc.owncloud.org/server/10.0/developer_manual/general/codingguidelines.html

File: tests/ui/features/lib/PersonalGeneralSettingsPage.php [hunk 3]
```diff
@@ -75,4 +79,37 @@ public function waitTillPageIsLoaded(
...
+	 * @param string $newPassword
+	 *
+	 * @return void
+	 */
+	public function changePassword($oldPassword, $newPassword) {
+		$this->fillField($this->newPasswordInputID, $newPassword);
+		$this->fillField($this->oldPasswordInputID, $oldPassword);
+		$changePasswordButton = $this->findById($this->changePasswordButtonID);
+		if (is_null($changePasswordButton)) {
+			throw new ElementNotFoundException(__METHOD__ . " could not find element with id $this->changePasswordButtonID");
```
Comment (individual-it) [inline tests/ui/features/lib/PersonalGeneralSettingsPage.php]: too long line

File: tests/ui/features/lib/PersonalGeneralSettingsPage.php [hunk 4]
```diff
@@ -75,4 +79,37 @@ public function waitTillPageIsLoaded(
...
+	 * @throws ElementNotFoundException
+	 *
+	 * @param Session $session
+	 * @return string
+	 */
+	public function getWrongPasswordMessageText(Session $session) {
+		$errorMessage = $this->findById($this->passwordErrorMessageID);
+		
+		if (is_null($errorMessage)) {
+			throw new ElementNotFoundException(__METHOD__ . " could not find element with id $this->passwordErrorMessageID");
```
Comment (individual-it) [inline tests/ui/features/lib/PersonalGeneralSettingsPage.php]: too long line

File: tests/ui/features/lib/PersonalGeneralSettingsPage.php [hunk 5]
```diff
@@ -75,4 +79,37 @@ public function waitTillPageIsLoaded(
...
+		}
+		$this->waitForAjaxCallsToStartAndFinish($session);
```
Comment (individual-it) [inline tests/ui/features/lib/PersonalGeneralSettingsPage.php]: I don't think that waiting makes sense here. No Ajax request should be active here, you have already got all information and haven't done any action (like click).
but at the end of `public function changePassword($oldPassword, $newPassword) {` that might be very usefull

File: tests/ui/features/other/changePasswordFromSetting.feature [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+		And I go to the personal general settings page
+
+	Scenario: Change password from setting
+		When I change the password to "passcode"
+		And I relogin with username "user1" and password "passcode"
+		Then I should be redirected to a page with the title "Files - ownCloud"
+
+	Scenario: Password change with wrong current password
+		When I change the password to "passcode" using wrong current password
+		Then a message should be displayed with the text "Wrong password"
```
Comment (individual-it) [inline tests/ui/features/other/changePasswordFromSetting.feature]: I think we should have a more specific step here. Something like `a password error message should be displayed ...`

**Review by individual-it — CHANGES_REQUESTED** [review]

File: tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php [hunk 7]
```diff
@@ -77,4 +82,62 @@ public function iChangeTheLanguageTo($language) {
...
+	 *
+	 * @param string $newPassword
+	 *
+	 * @return void
+	 */
+	public function iChangeThePasswordTo($newPassword) {
+		$username = $this->featureContext->getCurrentUser();
+		$oldPassword = trim($this->featureContext->getUserPassword($username));
+		$this->personalGeneralSettingsPage->changePassword($oldPassword, $newPassword,
+		$this->getSession());
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php]: Format - if you need to use extra lines, then the method/function name and ``(`` is on line1, then a line (or lines) for the parameters, then the closing ``)`` on its own line. Something like:
```
$this->personalGeneralSettingsPage->changePassword(
    $oldPassword, $newPassword, $this->getSession()
);
```

File: tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php [hunk 8]
```diff
@@ -77,4 +82,62 @@ public function iChangeTheLanguageTo($language) {
...
+	 *
+	 * @param string $newPassword
+	 *
+	 * @return void
+	 */
+	public function iChangeThePasswordToUsingWrongCurrentPassword($newPassword) {
+		$username = $this->featureContext->getCurrentUser();
+		$oldPassword = "thisisawrongpassword";
+		$this->personalGeneralSettingsPage->changePassword($oldPassword, $newPassword, 
+		$this->getSession());
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php]: Format here also

File: tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php [hunk 9]
```diff
@@ -77,4 +82,62 @@ public function iChangeTheLanguageTo($language) {
...
+	}
+	
+	/**
+	 * @Then a password error message should be displayed with the text :wrongPasswordmessageText
+	 *
+	 * @param string $wrongPasswordmessageText
+	 *
+	 * @return void
+	 */
+	public function aMessageShouldBeDisplayedWithTheText(
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php]: This function name might get duplicated by a general function checking for a message.
Maybe: ``aPasswordErrprMessageShouldBeDisplayedWithTheText``

File: tests/ui/features/lib/PersonalGeneralSettingsPage.php [hunk 10]
```diff
@@ -44,7 +49,6 @@ class PersonalGeneralSettingsPage extends OwncloudPage {
 	public function changeLanguage($language) {
 		$this->selectFieldOption($this->languageSelectId, $language);
 	}
-
```
Comment (phil-davis) [inline tests/ui/features/lib/PersonalGeneralSettingsPage.php]: While editing, try not to remove blank lines between functions

File: tests/ui/features/other/changePasswordFromSetting.feature [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+	Background:
+		Given these users exist:
+		|username|password|displayname|email       |
+		|user1   |1234    |User One   |u1@oc.com.np|
+		And I am on the login page
+		And I login with username "user1" and password "1234"
+		And I go to the personal general settings page
+
+	Scenario: Change password from setting
```
Comment (phil-davis) [inline tests/ui/features/other/changePasswordFromSetting.feature]: I am not sure what "from setting" means here. This ordinary scenario could just be called
```
Scenario: Change password
```

**Review by phil-davis — CHANGES_REQUESTED** [review]
I made a few comments about formatting and words.

**Comment by sushilkshetri** [comment]
@phil-davis  @individual-it  please review again.

File: tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php [hunk 12]
```diff
@@ -77,4 +82,62 @@ public function iChangeTheLanguageTo($language) {
...
+	
+	/**
+	 * @When I change the password to :newPassword using wrong current password
+	 *
+	 * @param string $newPassword
+	 *
+	 * @return void
+	 */
+	public function iChangeThePasswordToUsingWrongCurrentPassword($newPassword) {
+		$username = $this->featureContext->getCurrentUser();
```
Comment (phil-davis) [inline tests/ui/features/bootstrap/PersonalGeneralSettingsContext.php]: ``$username`` is unused here - the line can be deleted

**Review by phil-davis — COMMENTED** [review]
Just 1 unused var that I can see.

**Comment by sushilkshetri** [comment]
@phil-davis  @individual-it  please review again.

**Review by phil-davis — APPROVED** [review]
LGTM

**Comment by individual-it** [comment]
please rebase to make the CI run again

**Comment by individual-it** [comment]
@ownclouders rebase

**Comment by ownclouders** [comment]
Hey! I'm [GitMate.io](https://gitmate.io)! This pull request is being rebased automatically. Please **DO NOT** push while rebase is in progress or your changes would be lost permanently :warning:

**Comment by ownclouders** [comment]
Automated rebase with [GitMate.io](https://gitmate.io) was successful! :tada:

**Comment by individual-it** [comment]
rebased by using the bot

**Review by individual-it — APPROVED** [review]

**Comment by individual-it** [comment]
backport in #30545

**Pull request merged**